tests/federation_keys_test.go

```diff
@@ -29,60 +29,226 @@ import (
+			}),
+			match.JSONMapEach("old_verify_keys", func(k, v gjson.Result) error {
+				// Check for ed25519 prefix
+				if !strings.HasPrefix(k.Str, "ed25519:") {
+					return fmt.Errorf("old_verify_keys: Key '%s' has no 'ed25519:' prefix", k.Str)
+				}
+
+				expiredTs := v.Get("expired_ts")
+
+				// Test expired_ts existence and number type
+				if !expiredTs.Exists() {
+					return fmt.Errorf("old_verify_keys: Key '%s' has no expired_ts in value", k.Str)
+				}
+				if expiredTs.Type != gjson.Number {
+					return fmt.Errorf("old_verify_keys: Key '%s' has expired_ts with unexpected type, expected Number, got '%s'", k.Str, expiredTs.Type.String())
+				}
+
+				key := v.Get("key")
+
+				// Test key existence and string type
+				if !key.Exists() {
+					return fmt.Errorf("old_verify_keys: Key '%s' has no expired_ts in value", k.Str)
+				}
+				if key.Type != gjson.String {
+					return fmt.Errorf("old_verify_keys: Key '%s' has expired_ts with unexpected type, expected String, got '%s'", k.Str, key.Type.String())
 				}
+
 				var keyBytes []byte
-				keyBytes, err = base64.RawStdEncoding.DecodeString(v.Get("key").Str)
+				keyBytes, err = base64.RawStdEncoding.DecodeString(key.Str)
 				if err != nil {
-					return fmt.Errorf("Failed to decode ed25519 key as base64: %s", err)
+					return fmt.Errorf("old_verify_keys: Key '%s' failed to decode as ed25519 base64: %s", k.Str, err)
+				}
+
+				oldKeys[k.Str] = keyBytes
+
+				return nil
+			}),
+
+			// Check signatures
+			match.JSONMapEach("signatures", func(k, v gjson.Result) error {
+				if !v.IsObject() {
+					return fmt.Errorf("signatures: Value for server '%s' is not an object", k.Str)
+				}
+
+				for key, v := range v.Map() {
+					// Check for ed25519 prefix
+					if !strings.HasPrefix(key, "ed25519:") {
+						return fmt.Errorf("signatures: Key '%s' for server '%s' has no 'ed25519:' prefix", key, k.Str)
+					}
+
+					if v.Type != gjson.String {
+						return fmt.Errorf("signatures: Key '%s' for server '%s' has unexpected type, expected String, got '%s'", key, k.Str, v.Type.String())
+					}
+
+					_, err = base64.RawStdEncoding.DecodeString(v.Str)
+					if err != nil {
+						return fmt.Errorf("old_verify_keys: Key '%s' failed to decode as ed25519 base64: %s", k.Str, err)
+					}
 				}
-				// slightly evil to side-effect like this, but it's the easiest way
-				keyID = k.Str
-				key = keyBytes
+
 				return nil
 			}),
 		},
 	})
+
 	jsonObj := gjson.ParseBytes(body)
 	gotTime := time.Unix(0, jsonObj.Get("valid_until_ts").Int()*1000*1000) // ms -> ns
 	wantTime := time.Now()
 	if gotTime.Before(wantTime) {
-		t.Errorf("valid_until_ts : timestamp is in the past: %s < %s", gotTime, wantTime)
+		t.Errorf("valid_until_ts: timestamp is in the past: %s < %s", gotTime, wantTime)
+	}
+
+	checkKeysAndSignatures(t, body, jsonObj, keys, oldKeys)
+}
+
+func checkKeysAndSignatures(t *testing.T, body []byte, jsonObj gjson.Result, keys, oldKeys map[string]ed25519.PublicKey) {
```

Comment:
It seems needless to make an additional function here given it's only used in this one test, and makes assumptions like it being `hs1`. I would just inline this function into the test personally.

Reply:
I'm not sure if that's possible, the CI Linter complained about complexity :(